presto-main/src/main/java/com/facebook/presto/dispatcher/QueryInfoFetcher.java

```diff
@@ -0,0 +1,323 @@
+import com.facebook.airlift.http.client.Request;
+import com.facebook.airlift.http.client.ResponseHandler;
+import com.facebook.presto.execution.LocationFactory;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.metadata.InternalNode;
+import com.facebook.presto.server.BasicQueryInfo;
+import com.facebook.presto.server.RequestErrorTracker;
+import com.facebook.presto.server.SimpleHttpResponseCallback;
+import com.facebook.presto.server.SimpleHttpResponseHandler;
+import com.facebook.presto.server.smile.BaseResponse;
+import com.facebook.presto.server.smile.Codec;
+import com.facebook.presto.server.smile.SmileCodec;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.Duration;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.ws.rs.core.UriBuilder;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.facebook.airlift.http.client.Request.Builder.prepareGet;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CURRENT_STATE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_MAX_WAIT;
+import static com.facebook.presto.server.RequestErrorTracker.queryRequestErrorTracker;
+import static com.facebook.presto.server.RequestHelpers.setContentTypeHeaders;
+import static com.facebook.presto.server.smile.AdaptingJsonResponseHandler.createAdaptingJsonResponseHandler;
+import static com.facebook.presto.server.smile.FullSmileResponseHandler.createFullSmileResponseHandler;
+import static com.facebook.presto.server.smile.JsonCodecWrapper.unwrapJsonCodec;
+import static com.facebook.presto.spi.StandardErrorCode.REMOTE_QUERY_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.units.Duration.nanosSince;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class QueryInfoFetcher
+        implements SimpleHttpResponseCallback<BasicQueryInfo>
+{
+    private final QueryId queryId;
+    private final Consumer<Throwable> onFail;
+    private final StateMachine<BasicQueryInfo> queryInfo;
+    private final StateMachine<Optional<BasicQueryInfo>> finalQueryInfo;
+    private final Codec<BasicQueryInfo> queryInfoCodec;
+
+    private final long updateIntervalMillis;
+    private final Duration queryInfoRefreshMaxWait;
+    private final AtomicLong lastUpdateNanos = new AtomicLong();
+    private final Duration maxErrorDuration;
+
+    private final HttpClient httpClient;
```

Comment:
This reminds me the thrift protocol revamp effort. Likely this will be refactored soon. Maybe add a TODO to indicate that. Ideally we will have something like the ones for `RpcShuffleClient`

Reply:
I think for the near future, we'll continue to use these `HttpClient`s, as we're just changing the serialization.